recover/recover.c

```diff
@@ -0,0 +1,78 @@
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+// Block size for memory card
+static const int BLOCK_SIZE = 512;
+
+// Filename length
+static const char *FILENAME_FORMAT = "###.jpg";
+
+// Prototypes
+bool jpg_signature_found(uint8_t buffer[BLOCK_SIZE]);
+FILE *create_jpg_file(int counter);
+
+int main(int argc, char *argv[])
+{
+    // Check for invalid usage
+    if (argc != 2)
+    {
+        printf("Usage: ./recover [filepath]\n");
+        return 1;
+    }
+
+    char *infile = argv[1];
+
+    FILE *card = fopen(infile, "r");
+    if (card == NULL)
+    {
+        printf("Could not open %s.\n", infile);
+        return 2;
+    }
+
+    // Create a buffer to look for jpeg signatures
+    uint8_t buffer[BLOCK_SIZE];
+    int counter = 0;
+
+    FILE *file = create_jpg_file(counter);
+
+    while (fread(buffer, 1, BLOCK_SIZE, card) == BLOCK_SIZE)
+    {
+        if (jpg_signature_found(buffer))
+        {
+            if (file != NULL)
+            {
+                fclose(file);
+            }
+            file = create_jpg_file(counter);
+            counter++;
+        }
+        fwrite(buffer, 1, BLOCK_SIZE, file);
```

Comment:
What would I assert here and how would I validate it?

Reply:
The short answer is something like 
```suggestion
        const int bytes_written = fwrite(buffer, 1, BLOCK_SIZE, file);
        assert(bytes_written == BLOCK_SIZE);
```

The longer answer is that the man pages will tell you what the return value of all these functions is. For convenience, [here's the man page for `fwrite()`](https://man7.org/linux/man-pages/man3/fwrite.3p.html). If you look at the section on the return value, you'll see that it returns the number of elements that were written. The number of elements that you are trying to write is the 3rd parameter of `fwrite()`. Based on this, you know that everything was written if the return value is the same as `nitems`.

Reply:
`assert` is what I was looking for! Another `if` loop stops the buffer from updating unless I stick another `while` loop inside it.